### PR TITLE
New zabbix_user_group module

### PR DESCRIPTION
**Pull request by emriver**
##### SUMMARY

New module to manage Zabbix user groups.
Tested on Zabbix 4.0.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

zabbix_user_group

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (rubentsirunyan) [inline lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py]: Typo here.
```suggestion
                options=dict(
```

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (rubentsirunyan) [inline lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py]: Consider using the common helper for missing dependency errors.
You can refer to this code as an example.

https://github.com/ansible/ansible/blob/42a831d42c17ee7d5226796fcdcac0d6fd69b1ba/lib/ansible/modules/monitoring/zabbix/zabbix_mediatype.py#L545-L546

**Review by rubentsirunyan — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (rubentsirunyan) [inline lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py:57]: I would prefer the term `permissions` for this option as it is called so in both Zabbix front end and API. On the other hand `permissions['permission']`  can be misleading.

Comment (emriver) [reply lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py:57]: I may have misunderstood but it seems to be called `rights` at least in the API  (https://www.zabbix.com/documentation/4.0/manual/api/reference/usergroup/create)

    Parameter | Type | Description
    rights | object/array | Permissions to assign to the group

Comment (rubentsirunyan) [reply lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py:57]: Yes, I think you're right. I just looked at [this](https://www.zabbix.com/documentation/4.0/manual/api/reference/usergroup/object#permission) section.

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (emriver) [reply lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py]: Fixed in the last commit

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (emriver) [reply lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py]: Fixed in the last commit

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (D3DeFi) [inline lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py:34]: If I am reading it correctly, `debug_mode` is a specific thing for a whole Zabbix API and not just for this module. 

This is not something we use in other zabbix modules and its usefulness for lets say an automated playbook is questionable. From the point of user's view, I shouldn't be expected to debug anything if only thing I am interested in is to create a few user groups.

Can you provide some reasoning behind this please? I can understand that it is a good help for development purposes.

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (D3DeFi) [inline lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py]: Using `name` as an identifier for users can lead to a lots of confusion. Imagine that I have three users named John but with different aliases.

Also this makes users without `name` attribute (service users or guest account) "invisible" for this module. Please, rework this to use `alias` instead as [Zabbix API documentation](https://www.zabbix.com/documentation/4.0/manual/api/reference/user/object) states that it is the required attribute for user object.

Comment (D3DeFi) [inline lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py]: Using `name` as an identifier for users can lead to a lots of confusion. Imagine that I have three users named John but with different aliases.

Also this makes users without `name` attribute (service users or guest account) "invisible" for this module. Please, rework this to use `alias` instead as [Zabbix API documentation](https://www.zabbix.com/documentation/4.0/manual/api/reference/user/object) states that it is the required attribute for user object.

**Review by D3DeFi — CHANGES_REQUESTED** [review]
Hello and thank you for the contribution @emriver ! I've tested this module with almost every currently supported zabbix version and it seems to be working properly, except a small inconsistency with identifying users (see my comments in the review).

After required adjustments are made, I think this module will be ready to be merged.

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (D3DeFi) [reply lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py:57]: @rubentsirunyan thanks a lot for taking a look at this! :)

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (emriver) [reply lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py:34]: I think there is a misunderstanding with this property.
From my understanding the `debug_mode`, (described [here](https://www.zabbix.com/documentation/4.0/manual/web_interface/debug_mode)) allows Zabbix users to click on a `Debug` button to troubleshoot front end web pages.

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (emriver) [reply lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py]: Good catch! I will update the code as requested

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (D3DeFi) [reply lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py:34]: Ok, my bad. Thanks for clarification! :)

**Review by D3DeFi — APPROVED** [review]
LGTM

**Review by rubentsirunyan — APPROVED** [review]
shipit

**Comment by D3DeFi** [comment]
Thanks for your work @emriver . This should now get merged soon, by someone with merge permissions (new modules are not auto-merged by default if I am not mistaken)

@rubentsirunyan thanks for the review

**Comment by rubentsirunyan** [comment]
I am testing this module against Zabbix 3.4 and it seems that the only issue is with `gui_access` argument. They changed it from this

>  0 - (default) use the system default authentication method;
1 - use internal authentication;
2 - disable access to the frontend. 

to this

> 0 - (default) use the system default authentication method;
1 - use internal authentication;
2 - use LDAP authentication;
3 - disable access to the frontend.  

So I would suggest adding something like this
```python
if float(self._zapi.api_version().rsplit('.', 1)[0]) >= 4.0:
    gui_accesses = {'default': '0', 'internal': '1', 'ldap': '2', 'disabled': '3'}
else: 
    gui_acesses = {'default': '0', 'internal': '1', 'disabled': '2'}
```
here:
https://github.com/ansible/ansible/blob/9ab4654c3542fff784bf0c0f032fa867091216a5/lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py#L190-L191

**Comment by rubentsirunyan** [comment]
Another useful change for me would be changing the type of `rights.host_group` from `str` to `list` (and maybe rename it to `host_groups` :) ) and let users set permissions for a group of host_groups and not for each host_group separately, as setting permissions for every single group (especially if you manage hundreds) is not going to be a fun job to do.
This will also allow us to add a keyword `all` for `host_groups` as some kind of "syntactic sugar" to query all host groups available.
I tweaked your section for generating the `rights` a little bit and transformed it to this function:
```python
def generate_rights_config(self, rights):
    generated_rights = []
    for right in rights:
        if right['host_groups'] == ['all']:
            groups = self._zapi.hostgroup.get({'output': 'extend'})
        else:
            groups = self._zapi.hostgroup.get({'filter': {'name': right['host_groups']}})
        if not groups:
            self._module.fail_json(msg="No host group %s was found" % right['host_groups'])
        permission = '0'
        if right['permission'] == 'RO':
            permission = '2'
        elif right['permission'] == 'RW':
            permission = '3'
        for group in groups: 
            group_id = group['groupid']
            generated_rights.append({'id': group_id, 'permission': permission})
    return generated_rights
```
BTW, this is just a "works for me" thing still :)

**Comment by rubentsirunyan** [comment]
@D3DeFi, @emriver, what do you think?

**Comment by D3DeFi** [comment]
> 
> 
> I am testing this module against Zabbix 3.4 and it seems that the only issue is with `gui_access` argument. They changed it from this
> 
> > 0 - (default) use the system default authentication method;
> > 1 - use internal authentication;
> > 2 - disable access to the frontend.
> 
> to this
> 
> > 0 - (default) use the system default authentication method;
> > 1 - use internal authentication;
> > 2 - use LDAP authentication;
> > 3 - disable access to the frontend.
> 
> So I would suggest adding something like this
> 
> ```python
> if float(self._zapi.api_version().rsplit('.', 1)[0]) >= 4.0:
>     gui_accesses = {'default': '0', 'internal': '1', 'ldap': '2', 'disabled': '3'}
> else: 
>     gui_acesses = {'default': '0', 'internal': '1', 'disabled': '2'}
> ```
> 
> here:
> 
> https://github.com/ansible/ansible/blob/9ab4654c3542fff784bf0c0f032fa867091216a5/lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py#L190-L191

Thanks for additional review @rubentsirunyan . As for comparing versions, there is [something similar in zabbix_template](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/monitoring/zabbix/zabbix_template.py#L351). I think it would be for the best to have it implemented in the same way for this module as well.

Regarding the other idea:  changing type of `rights.host_group` to `list`. This depends if @emriver has some time to implement this before this module gets merged. If not, @rubentsirunyan maybe you can add it later via PR as feature request.

**Comment by emriver** [comment]
> I am testing this module against Zabbix 3.4 and it seems that the only issue is with `gui_access` argument. They changed it from this
> 
> > 0 - (default) use the system default authentication method;
> > 1 - use internal authentication;
> > 2 - disable access to the frontend.
> 
> to this
> 
> > 0 - (default) use the system default authentication method;
> > 1 - use internal authentication;
> > 2 - use LDAP authentication;
> > 3 - disable access to the frontend.
> 
> So I would suggest adding something like this
> 
> ```python
> if float(self._zapi.api_version().rsplit('.', 1)[0]) >= 4.0:
>     gui_accesses = {'default': '0', 'internal': '1', 'ldap': '2', 'disabled': '3'}
> else: 
>     gui_acesses = {'default': '0', 'internal': '1', 'disabled': '2'}
> ```
> 
> here:
> 
> https://github.com/ansible/ansible/blob/9ab4654c3542fff784bf0c0f032fa867091216a5/lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py#L190-L191

Thanks for your review and tests @rubentsirunyan.
I'm ok with this improvement, should I push it now? or wait until this module is merged and then create a new pull request?

**Comment by emriver** [comment]
> Another useful change for me would be changing the type of `rights.host_group` from `str` to `list` (and maybe rename it to `host_groups` :) ) and let users set permissions for a group of host_groups and not for each host_group separately, as setting permissions for every single group (especially if you manage hundreds) is not going to be a fun job to do.
> This will also allow us to add a keyword `all` for `host_groups` as some kind of "syntactic sugar" to query all host groups available.
> I tweaked your section for generating the `rights` a little bit and transformed it to this function:
> 
> ```python
> def generate_rights_config(self, rights):
>     generated_rights = []
>     for right in rights:
>         if right['host_groups'] == ['all']:
>             groups = self._zapi.hostgroup.get({'output': 'extend'})
>         else:
>             groups = self._zapi.hostgroup.get({'filter': {'name': right['host_groups']}})
>         if not groups:
>             self._module.fail_json(msg="No host group %s was found" % right['host_groups'])
>         permission = '0'
>         if right['permission'] == 'RO':
>             permission = '2'
>         elif right['permission'] == 'RW':
>             permission = '3'
>         for group in groups: 
>             group_id = group['groupid']
>             generated_rights.append({'id': group_id, 'permission': permission})
>     return generated_rights
> ```
> 
> BTW, this is just a "works for me" thing yet :)

That's a good point, it could be really useful in large environments.
Since it is your idea, I think you should add it via PR once this one is merged, what do you think?

Another idea for large environments would be to allow the usage of wild cards in group names by using the `searchWildcardsEnabled` option.

**Comment by D3DeFi** [comment]
@emriver, please fix `gui_access` in this PR. Support for multiple groups can be added later. I hope this module will be merged soon.

**Review by D3DeFi — APPROVED** [review]
LGTM

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (sky-joker) [inline lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py]:
```suggestion
version_added: "2.10"
```

The current devel version is 2.10, so I think `version_added` is `2.10`.
https://github.com/ansible/ansible/blob/devel/lib/ansible/release.py

Comment (emriver) [reply lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py]: Updated :)

Comment (sky-joker) [reply lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py]: Thanks for the quick fix :)

**Review by D3DeFi — APPROVED** [review]
To bad this wasn't merged to 2.9rc.
@rubentsirunyan are you ok with state of this module?

shipit

**Comment by rubentsirunyan** [comment]
Yes, I'll create an issue with my suggestion above so that we don't forget about it in the future.
shipit

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (resmo) [inline lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py]: We used to allow a "dump" state in the past but this should functionality should go into an info e.g. `zabbix_user_group_info` module. Please remove this functionality from this module.

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (resmo) [inline lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py]: example should be removed.

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (resmo) [inline lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py]: remove dump state

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (resmo) [inline lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py]: remove dump state

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (resmo) [inline lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py]: update description, remove dump

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (resmo) [inline lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py]: update description, remove dump

**Review by resmo — CHANGES_REQUESTED** [review]
the state=dump should go into a dedicated info module

**Comment by richlv** [comment]
@emriver, sorry to bother you - have you had a chance to look at the suggested changes?
Seem to be simple enough before this could be merged.

**Review by emriver — COMMENTED** [review]
Ok, will update the PR

**Comment by emriver** [comment]
Hello !
Don't really know what to do here.
The zabbix documentation fragment specifies that `http_login_user` is required, but it is declared as `Required: False` in all Zabbix modules... and my build is failing because of that.

Should I:
- change the documentation fragment or remove the modules from the exception list?
- or add this module to the exception list?

**Comment by sky-joker** [comment]
Hi @emriver

The Zabbix module seems to write a target to ignore in `ignore.txt`
When I added `doc-required-mismatch` to `ignore.txt` and tested as follows, no error occurred.

```
# vi ./test/sanity/ignore.txt
(snip)
lib/ansible/modules/monitoring/zabbix/zabbix_user_info.py validate-modules:doc-required-mismatch
lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py validate-modules:doc-required-mismatch <- add
(snip)
```

@D3DeFi
Is my understanding this correct?

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (sky-joker) [inline lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py:305]:
```suggestion
        zbx.login(login_user, login_password)
        atexit.register(zbx.logout)
```

Please add `atexit.register(zbx.logout)`
see https://github.com/ansible/ansible/pull/58525

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (sky-joker) [inline lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py:114]:
```suggestion
import atexit
import traceback
```
Please import `atexit` module

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (sky-joker) [inline lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py]:
```suggestion
            - List of user aliases to add in the group
```

**Comment by D3DeFi** [comment]
> 
> 
> Hi @emriver
> 
> The Zabbix module seems to write a target to ignore in `ignore.txt`
> When I added `doc-required-mismatch` to `ignore.txt` and tested as follows, no error occurred.
> 
> ```
> # vi ./test/sanity/ignore.txt
> (snip)
> lib/ansible/modules/monitoring/zabbix/zabbix_user_info.py validate-modules:doc-required-mismatch
> lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py validate-modules:doc-required-mismatch <- add
> (snip)
> ```
> 
> @D3DeFi
> Is my understanding this correct?
> #65437

Seems about right, but I am leaning more towards fixing this properly. As @emriver pointed out, only inconsistency seems to be with `http_login_user`, which imo shouldn't be required. I've opened a new PR for this as it shouldn't be pushed alongside a new module in a same PR.

Thanks for discovering this guys. Until it gets merged, please implement @sky-joker 's suggestions.

File: lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py
Comment (D3DeFi) [inline lib/ansible/modules/monitoring/zabbix/zabbix_user_group.py]:
```suggestion
    - "zabbix-api >= 0.5.4"
```
as `zbx.logout()` wasn't present in `0.5.3`.

File: lib/ansible/plugins/doc_fragments/zabbix.py
Comment (D3DeFi) [inline lib/ansible/plugins/doc_fragments/zabbix.py]: remove this file from the PR altogether please

**Review by D3DeFi — APPROVED** [review]
shipit

File: test/sanity/ignore.txt
Comment (mattclay) [inline test/sanity/ignore.txt]: New ignores should not be added. Correct the issue with the module instead. If the issue is in a shared doc fragment, fix that in a separate PR first.

Comment (D3DeFi) [reply test/sanity/ignore.txt]: @mattclay #65631

Comment (mattclay) [reply test/sanity/ignore.txt]: Merged. Thanks.

**Comment by D3DeFi** [comment]
Looks good to me now. Can we merge @mattclay ?

**Comment by juanluisbaptiste** [comment]
Hi,

What's the state of this module ? will be merged soon ?

Thanks.

**Comment by sky-joker** [comment]
Hi @emriver

The Zabbix modules moved to collections repository, so can you please add an integration test and move this PR to this [repo](https://github.com/ansible-collections/community.zabbix)?

**Comment by D3DeFi** [comment]
Thanks @sky-joker for explaining the situation. 

Merge of this module unfortunately took so long that the ansible is no longer accepting any community modules. Everything is being migrated to collections (which will be shipped with ansible 2.10 release).

On the bright side, we now would be able to merge your module quicker than before :) Sorry for the inconvenience.

**Comment by D3DeFi** [comment]
I know we mentioned this earlier, but here is comment that we are pasting to each open PR regarding Zabbix comment:

Unfortunately, your contribution can no longer be merged to [ansible/ansible](https://github.com/ansible/ansible). The way how community content is being shipped within Ansible is [changing](https://docs.ansible.com/ansible/devel/user_guide/collections_using.html#collections) as of right now with the preparations for Ansible 2.10. 

All development regarding Zabbix content has been moved to a separate [ansible-collections/community.zabbix](https://github.com/ansible-collections/community.zabbix) repository. Please resubmit your PR against the new repository with the changes we have requested in the previous comments and reviews (if any).

This separation now allows us to address your contribution much quicker than before. We are giving each contributor a few weeks (roughly a month) notice to move his/hers PRs to a new repository.

After a month, we will try to migrate all leftover PRs by ourselves, but author information may get lost in this way.

Sorry for the inconvenience and thank you for the understanding.

close_me

**Pull request closed without merging**
